### PR TITLE
MultivariateIntPolynomial

### PR DESCRIPTION
**Pull request by chenchfort**
This code currently does not build.  Currently compilation stops at

[  7%] Building CXX object symengine/CMakeFiles/symengine.dir/basic.cpp.o

with the errors

In file included from /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/basic.h:36:0,
                 from /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/basic.cpp:1:
                 /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/dict.h:30:38: error: invalid declarator before Ã¢umap_basic_numÃ¢
                          RCPBasicHash, RCPBasicKeyEq> umap_basic_num;
                                                                ^
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/dict.h:202:62: error: Ã¢umap_basic_numÃ¢ in namespace Ã¢SymEngineÃ¢ does not name a type
 std::ostream& operator<<(std::ostream& out, const SymEngine::umap_basic_num& d);
                                                               ^
In file included from /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/basic.cpp:3:0:
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/add.h:18:5: error: Ã¢umap_basic_numÃ¢ does not name a type
     umap_basic_num dict_; //! The dictionary of the rest (e.g. `x+y` in `2+x+y`)
     ^
In file included from /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/basic.cpp:3:0:
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/add.h:25:40: error: Ã¢umap_basic_numÃ¢ has not been declared
     Add(const RCP<const Number> &coef, umap_basic_num&& dict);
                                        ^
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/add.h:41:70: error: Ã¢umap_basic_numÃ¢ has not been declared
     static RCP<const Basic> from_dict(const RCP<const Number> &coef, umap_basic_num &&d);
                                                                      ^
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/add.h:45:31: error: Ã¢umap_basic_numÃ¢ has not been declared
     static void dict_add_term(umap_basic_num &d,
                                    ^
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/add.h:51:72: error: Ã¢umap_basic_numÃ¢ has not been declared
     static void coef_dict_add_term(const Ptr<RCP<const Number>> &coef, umap_basic_num &d,
                                                                             ^
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/add.h:61:19: error: Ã¢umap_basic_numÃ¢ does not name a type
             const umap_basic_num& dict) const;
                                ^
In file included from /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/printer.h:4:0,
                 from /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/basic.cpp:12:
                 /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/visitor.h: In member function Ã¢void SymEngine::CoeffVisitor::bvisit(const SymEngine::Add&)Ã¢:
                 /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/visitor.h:94:9: error: Ã¢umap_basic_numÃ¢ was not declared in this scope
                          umap_basic_num dict;
                 /home/myluszcz/ECS193/SYMENGINE/symengine/symengine/visitor.h:94:9: error: Ã¢umap_basic_numÃ¢ was not declared in this scope
                          umap_basic_num dict;
                                   ^
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/visitor.h:96:25: error: Ã¢const class SymEngine::AddÃ¢ has no member named Ã¢dict_Ã¢
         for (auto &p: x.dict_) {
                                  ^
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/visitor.h:99:55: error: Ã¢dictÃ¢ was not declared in this scope
                 Add::coef_dict_add_term(outArg(coef), dict, p.second, coeff_);
                                                                        ^
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/visitor.h:102:49: error: Ã¢dictÃ¢ was not declared in this scope
         coeff_ = Add::from_dict(coef, std::move(dict));
                                                          ^
symengine/CMakeFiles/symengine.dir/build.make:86: recipe for target 'symengine/CMakeFiles/symengine.dir/basic.cpp.o' failed
make[2]: *** [symengine/CMakeFiles/symengine.dir/basic.cpp.o] Error 1
CMakeFiles/Makefile2:93: recipe for target 'symengine/CMakeFiles/symengine.dir/all' failed
make[1]: *** [symengine/CMakeFiles/symengine.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2

Our current goal WRT to this branch is to get the MultivariatePolynomial object to compile.  Once this occurs, we will write functions for performing addition, subtraction, negation, and multiplication, probably using the naive algorithms at first.  After this we will write testers and benchmarks against which to compare future improvements.

@shivamvats @certik @isuruf

### CODE REVIEW

**Comment by Sumith1896** [comment]
Let me know if you need any help on this. I'll have a look at this later
too.

**Comment by rwst** [comment]
Semicolon missng in line 27 of symengine/dict.h

**Comment by myluszczak** [comment]
The code from the most recent commit (d884c7b500ad399f32bbb26dfffad0c7fdf646e2) now compiles on my computer, but I had to comment out the contents of test_series_generic.cpp and series.cpp in /benchmaks.  Those files test code that Iris and Charles are writing, and their most recent work is not yet merged into this branch.  

Github tells me that it's been unable to build on appveyor, though.  Right now it won't let me see the terminal output, but gives me the message "Failure: AppVeyor wars unable to build non-mergeable pull request."

mul_mult_poly still needs to be implemented.

@rwst @certik @isuruf @shivamvats @Sumith1896

**Comment by shivamvats** [comment]
@myluszczak You need to merge the master branch into this one and then resolve the merge conflicts.

**Comment by myluszczak** [comment]
@#shivamvats Thanks.

**Comment by rwst** [comment]
The branch has merge conflicts. Please resolve.

File: symengine/dict.cpp [hunk 1]
```diff
@@ -190,43 +190,51 @@ int map_int_Expr_compare(const map_int_Expr &A, const map_int_Expr &B)
...
-int multiset_basic_compare(const multiset_basic &a, const multiset_basic &b)
-{
-    return set_compare<multiset_basic>(a, b);
+int umap_uvec_mpz_compare(const umap_uvec_mpz &a, const umap_uvec_mpz &b){
+    if(a.size() < b.size())
+        return (a.size() < b.size()) ? -1 : 1;
+    return 0;
+}
+
+//coppied from umap_eq, with derefrencing of image in map removed.
```
Comment (shivamvats) [inline symengine/dict.cpp]: *Copied

**Comment by shivamvats** [comment]
@myluszczak Most of the operations seem to have been implemented.  What else do you want to do here?
Btw, James' commits don't seem to be linked to his github account.

**Comment by myluszczak** [comment]
@shivamvats Well, so far as MultivariateIntPolynomial, the only major thing I wanted to do was add some more tests to improve code coverage.  After that, we'll need to make a polynomial that can take expression coefficients, and then an object with overloaded operators, but that should be fairly simple.

I didn't notice that Jame's commits were not linked to his account.

**Comment by StoicBronco** [comment]
How would I link my account to my commits? Git is not my strongest suit :/

**Comment by myluszczak** [comment]
@StoicBronco I'm not sure what could be causing that.  When you set your email on your local repository, is it the same email associated with your GitHub account?

**Comment by shivamvats** [comment]
@myluszczak Cool! Lets get this PR merged first. You can send another PR for expression coefficients and overloaded operators.
@StoicBronco Looks like the commits are linked now.

**Comment by myluszczak** [comment]
@shivamvats Alright.  Should we rebase now, or is there anything else that should be added to this PR?

**Comment by chenchfort** [comment]
Some commits in the history are irrelevant now. Might need to figure out how to get rid of them.

**Comment by myluszczak** [comment]
@chenchfort Which commits were you thinking of?

**Comment by chenchfort** [comment]
The ones that have already been merged into master.

**Comment by myluszczak** [comment]
@chenchfort I think those are just yours and Iris's, which I brought in when I merged with your branch.  Any others?

**Comment by chenchfort** [comment]
Let's see what @isuruf has to say.

**Comment by isuruf** [comment]
Yes, most of @chenchfort's and @irislq's commits were merged and we need to figure out a way to remove those from this PR.

**Comment by chenchfort** [comment]
Will rebase do it?

**Comment by myluszczak** [comment]
We could squash them before rebasing

**Comment by isuruf** [comment]
Anything that works

**Comment by shivamvats** [comment]
Yes, you could at least squash the merge commits. Rebasing should take care of the other commits already in master.

**Comment by myluszczak** [comment]
Alright.  If there's nothing else, I'll start rebasing now.

**Comment by myluszczak** [comment]
@chenchfort is the current master of our fork up to date?

**Comment by chenchfort** [comment]
@myluszczak ok done. Can always pull from upstream.

**Comment by myluszczak** [comment]
After 11 hours, I have been unable to rebase effectively.  I'm doing something wrong but I don't know what.  I'm going to go to sleep and will probably not be able to work or be in contact for the rest of the day

**Comment by myluszczak** [comment]
I think that I am going to have to have someone walk me through rebasing.  I don't have any idea what I'm doing.  When I don't use any merge strategies, I get merge conflicts and introduce errors resolving them, and when I use merge strategies the result does not compile.  I am going to need someone to hold my hand and walk me through the process.  If someone can help, that would be appreciated.

**Comment by shivamvats** [comment]
@myluszczak Could you confirm that the master is now merged with this branch? I think Ondrej did that for you or was it you? 

Now what you have are mostly redundant merge commits which you can squash if you want to.

**Comment by myluszczak** [comment]
@shivamvats How would you confirm that mater is merged?

**Comment by shivamvats** [comment]
If you look at the commits in this PR, the redundant ones are no more there (there were more than 100 earlier IIRC). Are there other commits that shouldn't be here?

**Comment by myluszczak** [comment]
Not so far as I can tell, no.  There are a bunch of commits at the top corresponding to the merge, but those didn't show up in the interactive rebase; there were a bunch of other commits that I assume were brought in by these merges.

**Comment by shivamvats** [comment]
You could try using `--preserve-merges` option, though I haven't tried it before. @isuruf Suggestions?

**Comment by myluszczak** [comment]
@shivamvats So the overall command would be `git rebase -i --preserve-merges master`?  And then, in the interactive screen, squash 	ee78f3f - 	964752f ?

**Comment by shivamvats** [comment]
Yes. I hope it works.

**Comment by myluszczak** [comment]
Alright, I'll try it.

**Comment by shivamvats** [comment]
`git rebase --abort`
I'll look for a better way.

**Comment by myluszczak** [comment]
@shivamvats I tired that, but it tells me that there is no rebase in progress.  I'll just nuke my local repository and clone from the remote.

**Comment by shivamvats** [comment]
@myluszczak @StoicBronco This isn't rebased properly yet so that problem remains. I tried rebasing but there are too many conflicts. How many files have you changed? Could you copy the changes in another branch and send a fresh PR?

**Comment by myluszczak** [comment]
@shivamvats Yeah, copy the changes into another branch may be the best we can do at this point.  

Is there anything we should be doing to avoid this sort of situation in the future?  Merging more frequently, merging less frequently, rebasing instead of merging, rebasing then merging, that sort of thing?

**Comment by Sumith1896** [comment]
@myluszczak I don't exactly know what happened here the following workflow might help in future:

While implementing something new:
1) Update master and create a new branch out of it.
2) Keep adding your commits on it and push, it will be visible.
3) You can merge with master as and when the need be, for e.g., merge conflicts.

If there another member of your team who also wants to work on this, then:
1) After you add your commits, push
2) Request your teammates to send a PR with **their** commits to **your** branch.
3) Once something is merged, update your branches regularly.

**Comment by myluszczak** [comment]
@Sumith1896 Thanks for the advice.  With regards to cooperating through pull requests, what we've been doing is operating out of the same remote--We'd pull and resolve conflicts before beginning work and push when we were done.  Would that also work?

**Comment by myluszczak** [comment]
Work transferred to PR 891

**Comment by Sumith1896** [comment]
@myluszczak That should also work but you should be a bit careful then. As a thumb rule, branch as much as possible.

**Comment by Sumith1896** [comment]
Closing this as you moved your work.

**Comment by shivamvats** [comment]
@myluszczak I have just one point to add- avoid rebasing. It rewrites the commit history so you are better off merging. As for merging with master, you usually do it only if there has been some relevant change in the master that you want to use or if there are merge conflicts.

**Pull request closed without merging**